File: src/main/java/me/centralhardware/znatoki/telegram/statistic/telegram/Bot.java
package me.centralhardware.znatoki.telegram.statistic.telegram;

import jakarta.annotation.PostConstruct;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import me.centralhardware.telegram.bot.common.ClickhouseRuben;
import me.centralhardware.znatoki.telegram.statistic.Config;
import me.centralhardware.znatoki.telegram.statistic.entity.postgres.Role;
import me.centralhardware.znatoki.telegram.statistic.entity.postgres.TelegramUser;
import me.centralhardware.znatoki.telegram.statistic.mapper.postgres.UserMapper;
import me.centralhardware.znatoki.telegram.statistic.telegram.CommandHandler.CommandHandler;
import me.centralhardware.znatoki.telegram.statistic.telegram.callbackHandler.CallbackHandler;
import me.centralhardware.znatoki.telegram.statistic.telegram.fsm.ClientFsm;
import me.centralhardware.znatoki.telegram.statistic.telegram.fsm.PaymentFsm;
import me.centralhardware.znatoki.telegram.statistic.telegram.fsm.TimeFsm;
import org.springframework.stereotype.Component;
import org.telegram.telegrambots.client.okhttp.OkHttpTelegramClient;
import org.telegram.telegrambots.longpolling.interfaces.LongPollingUpdateConsumer;
import org.telegram.telegrambots.longpolling.starter.SpringLongPollingBot;
import org.telegram.telegrambots.meta.api.methods.commands.SetMyCommands;
import org.telegram.telegrambots.meta.api.objects.Message;
import org.telegram.telegrambots.meta.api.objects.Update;
import org.telegram.telegrambots.meta.api.objects.commands.BotCommand;

import java.util.List;
import java.util.Optional;

@Component
@Slf4j
@RequiredArgsConstructor
public class Bot implements SpringLongPollingBot {

    private final ClickhouseRuben clickhouse = new ClickhouseRuben();

    private final TelegramSender sender;
    private final TelegramUtil telegramUtil;
    private final List<CommandHandler> commandHandlers;
    private final List<CallbackHandler> callbackHandlers;
    private final InlineHandler inlineHandler;

    private final TimeFsm timeFsm;
    private final ClientFsm clientFsm;
    private final PaymentFsm paymentFsm;

    private final UserMapper userMapper;
    private final OkHttpTelegramClient telegramClient;

    @SneakyThrows
    @PostConstruct
    public void init() {
        var commands = SetMyCommands.builder()
                .commands(List.of(createCommand("/addtime", "Добавить запись"),
                        createCommand("/addpayment", "Добавить оплату"),
                        createCommand("/report", "Отчет за текущий месяц"),
                        createCommand("/reportprevious", "Отчет за предыдущий месяц"),
                        createCommand("/reset", "Сбросить состояние")))
                .build();
        telegramClient.execute(commands);
    }

    private BotCommand createCommand(String command, String description){
        return BotCommand
                .builder()
                .command(command)
                .description(description)
                .build();
    }

    @Override
    public LongPollingUpdateConsumer getUpdatesConsumer() {
        return updates -> updates.forEach(this::onUpdateReceived);
    }

    public void onUpdateReceived(Update update) {
        try {
            clickhouse.log(update, "znatokiStatistic");
            telegramUtil.logUpdate(update);

            Long userId = telegramUtil.getUserId(update);

            TelegramUser user = userMapper.getById(userId);

            if (user != null && user.getRole() == Role.BLOCK){
                log.info("Access blocked for user {}({})", user.getName(), user.getId());
                return;
            }

            boolean isStart = Optional.of(update)
                    .map(Update::getMessage)
                    .map(Message::getText)
                    .filter(it -> it.equals("/start"))
                    .isPresent();
            if (isStart) {
                processCommand(update);
                return;
            }

            if (userMapper.getById(userId) == null){
                sender.sendText("Вам необходимо создать или присоединиться к организации", TelegramUtil.getFrom(update));
                return;
            }

            if (processCommand(update)) return;

            if (inlineHandler.processInline(update)) return;

            if (processCallback(update)) return;

            if (timeFsm.isActive(userId)) {
                timeFsm.process(update);
                return;
            }

            if(clientFsm.isActive(userId)){
                clientFsm.process(update);
                return;
            }

            if(paymentFsm.isActive(userId)){
                paymentFsm.process(update);
                return;
            }

        } catch (Throwable t) {
            log.warn("Error while processing update", t);
        }
    }

    private boolean processCommand(Update update) {
        return processHandler(update, commandHandlers);
    }

    private boolean processCallback(Update update) {
        return processHandler(update, callbackHandlers);
    }

    private <T extends Handler> boolean processHandler(Update update, List<T> handlers){
        for (var handler : handlers) {
            if (handler.isAcceptable(update)) {
                handler.handle(update);
                return true;
            }
        }

        return false;
    }

    @Getter
    private final String botToken = Config.Telegram.getTelegramToken();

}

File: src/main/kotlin/me/centralhardware/znatoki/telegram/statistic/telegram/Bot.kt
package me.centralhardware.znatoki.telegram.statistic.telegram

import jakarta.annotation.PostConstruct
import me.centralhardware.telegram.bot.common.ClickhouseRuben
import me.centralhardware.znatoki.telegram.statistic.Config
import me.centralhardware.znatoki.telegram.statistic.entity.Role
import me.centralhardware.znatoki.telegram.statistic.mapper.UserMapper
import me.centralhardware.znatoki.telegram.statistic.telegram.callbackHandler.CallbackHandler
import me.centralhardware.znatoki.telegram.statistic.telegram.commandHandler.CommandHandler
import me.centralhardware.znatoki.telegram.statistic.telegram.fsm.Storage
import me.centralhardware.znatoki.telegram.statistic.userId
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.telegram.telegrambots.client.okhttp.OkHttpTelegramClient
import org.telegram.telegrambots.longpolling.interfaces.LongPollingUpdateConsumer
import org.telegram.telegrambots.longpolling.starter.SpringLongPollingBot
import org.telegram.telegrambots.meta.api.methods.commands.SetMyCommands
import org.telegram.telegrambots.meta.api.objects.Update
import org.telegram.telegrambots.meta.api.objects.commands.BotCommand

@Component
class Bot(
    private val sender: TelegramSender,
    private val commandHandlers: List<CommandHandler>,
    private val callbackHandlers: List<CallbackHandler>,
    private val inlineHandler: InlineHandler,
    private val userMapper: UserMapper,
    private val telegramClient: OkHttpTelegramClient,
    private val storage: Storage
) : SpringLongPollingBot {

    private val clickhouse = ClickhouseRuben()
    private val log: Logger = LoggerFactory.getLogger(Bot::class.java)

    @PostConstruct
    fun init() {
        val commands = SetMyCommands.builder()
            .commands(listOf(
                createCommand("/addtime", "Добавить запись"),
                createCommand("/addpayment", "Добавить оплату"),
                createCommand("/report", "Отчет за текущий месяц"),
                createCommand("/reportprevious", "Отчет за предыдущий месяц"),
                createCommand("/reset", "Сбросить состояние")))
            .build()
        telegramClient.execute(commands)
    }

    private fun createCommand(command: String, description: String): BotCommand =
        BotCommand.builder()
            .command(command)
            .description(description)
            .build()

    override fun getBotToken(): String = Config.Telegram.token

    override fun getUpdatesConsumer(): LongPollingUpdateConsumer =
        LongPollingUpdateConsumer { updates -> updates.forEach { onUpdateReceived(it) } }

    private fun onUpdateReceived(update: Update) {
        try {
            clickhouse.log(update, "znatokiStatistic")

            val userId = update.userId()

            val user = userMapper.getById(userId)

            user?.let {
                if (it.role == Role.BLOCK){
                    log.info("Access blocked for user ${user.name}(${user.id})")
                    return
                }
            }

            val isStart = update.message?.takeIf { it.text == "/start" } != null
            if (isStart) {
                processCommand(update)
                return
            }

            if (userMapper.getById(userId) == null){
                sender.sendText("Вам необходимо создать или присоединиться к организации", userId)
                return
            }

            if (processCommand(update)) return

            if (inlineHandler.processInline(update)) return

            if (processCallback(update)) return

            storage.takeIf { it.contain(userId) }?.process(userId, update)

        } catch (t: Throwable) {
            log.warn("Error while processing update", t)
        }
    }

    private fun processCommand(update: Update) = processHandler(update, commandHandlers)

    private fun processCallback(update: Update) = processHandler(update, callbackHandlers)

    private fun <T : Handler> processHandler(update: Update, handlers: List<T>): Boolean =
        handlers.find { handler -> handler.isAcceptable(update) }?.let {
            it.handle(update)
            true
        } ?: false
}